### PR TITLE
Reset the file to beginning when reading to the limitation for DSD file.

### LINKED ISSUE #32: Handle the end of file for DSD player
The DSD file player read the DSD file and play it. When it reaches the end of file, it is not currently handled well and causing underrun, which makes noise as it is a bit-perfect playback

### PR DESCRIPTION
**Pull request by flamme**
Fixes #32

### CODE REVIEW

File: app/src/main/java/com/google/android/soundchecker/dsd/DsfReader.kt [hunk 1]
```diff
@@ -147,9 +148,15 @@ class DsfReader(inputStream: InputStream) {
     private fun readNextBlock(channel: Int): Boolean {
         Arrays.fill(mData[channel], 0, mData[channel].size - 1, 0x0.toByte())
         try {
+            if (mDataLeft < mData[channel].size && mStream.markSupported()) {
+                Log.d(TAG, "Reach end of file, mark supported, reset to beginning")
```
Comment (robertwu1) [inline app/src/main/java/com/google/android/soundchecker/dsd/DsfReader.kt:152]: Add the word loop somewhere as it's not immediately obvious that the audio will continue to play after resetting to the beginning.

**Review by robertwu1 — APPROVED** [review]

File: app/src/main/java/com/google/android/soundchecker/dsd/DsfReader.kt [hunk 2]
```diff
@@ -147,9 +148,15 @@ class DsfReader(inputStream: InputStream) {
...
         try {
+            if (mDataLeft < mData[channel].size && mStream.markSupported()) {
+                Log.d(TAG, "Reach end of file, mark supported, reset and play from beginning")
+                mStream.reset()
+                mDataLeft = mDataTotal
+            }
             val sz = mStream.read(mData[channel])
             if (sz != mData[channel].size) {
                 Log.w(TAG, "Cannot read full buffer: $sz")
+                return false
```
Comment (philburk) [inline app/src/main/java/com/google/android/soundchecker/dsd/DsfReader.kt:159]: Do we want to support reading a partial buffer at the end?
This seems expected?
Maybe return false if nothing read. Or return size read.

File: app/src/main/java/com/google/android/soundchecker/dsd/DsfReader.kt [hunk 3]
```diff
@@ -147,9 +148,15 @@ class DsfReader(inputStream: InputStream) {
     private fun readNextBlock(channel: Int): Boolean {
         Arrays.fill(mData[channel], 0, mData[channel].size - 1, 0x0.toByte())
```
Comment (philburk) [inline app/src/main/java/com/google/android/soundchecker/dsd/DsfReader.kt:149]: Is 0x00 meant to be silence? Or is 0xAA silence?
Maybe if we pad partial buffers with DSD silence then we might not hear noise.

**Review by philburk — COMMENTED** [review]
reverseByte() might be simpler if it just used Int internally instead of casting back and forth Int<->Byte.
Is it used to fill a lookup table for fast real-time reverse?

File: app/src/main/java/com/google/android/soundchecker/dsd/DsfReader.kt (hunk 3 above)
Comment (flamme) [reply app/src/main/java/com/google/android/soundchecker/dsd/DsfReader.kt]: I think should be 0xa.

**Comment by flamme** [comment]
> reverseByte() might be simpler if it just used Int internally instead of casting back and forth Int<->Byte.
Is it used to fill a lookup table for fast real-time reverse?

If read into int, we need to take care of if the left data is not a times of 4 byte.

File: app/src/main/java/com/google/android/soundchecker/dsd/DsfReader.kt [hunk 4]
```diff
@@ -145,26 +145,27 @@ class DsfReader(inputStream: InputStream) {
...
     }
 
-    private fun readNextBlock(channel: Int): Boolean {
-        Arrays.fill(mData[channel], 0, mData[channel].size - 1, 0x0.toByte())
+    /**
+     * Returns the number of data read. Returns negative value if there is an error when reading or
+     * when it reaches the end of file and mark is not supported.
+     */
+    private fun readNextBlock(channel: Int): Int {
+        Arrays.fill(mData[channel], 0, mData[channel].size - 1, 0xA.toByte())
```
Comment (philburk) [inline app/src/main/java/com/google/android/soundchecker/dsd/DsfReader.kt:153]: 0xA == 0x0A
I think we need 0xAA for a full byte of DSD silence.

**Review by philburk — CHANGES_REQUESTED** [review]

**Comment by philburk** [comment]
> If read into int, we need to take care of if the left data is not a times of 4 byte.

We do not need to read into an Int. The method reverseByte() can take in a Byte and return a Byte.
But internally it can use Int. The CPU is 32 or 64 bit so it will be efficient and the code will be easier to read without all the casting.

Also is this used a lot at run-time? If so then it is common to use a reverse-byte lookup table.
This is a common technique for FFT implementations.

File: app/src/main/java/com/google/android/soundchecker/dsd/DsfReader.kt (hunk 4 above)
Comment (philburk) [reply app/src/main/java/com/google/android/soundchecker/dsd/DsfReader.kt:153]: Or 0x55 would also work and not have any sign extension issues.

Comment (flamme) [reply app/src/main/java/com/google/android/soundchecker/dsd/DsfReader.kt:153]: Per discussed, we should use 0x69 as silence.

**Comment by flamme** [comment]
> > If read into int, we need to take care of if the left data is not a times of 4 byte.
> 
> We do not need to read into an Int. The method reverseByte() can take in a Byte and return a Byte. But internally it can use Int. The CPU is 32 or 64 bit so it will be efficient and the code will be easier to read without all the casting.
> 
> Also is this used a lot at run-time? If so then it is common to use a reverse-byte lookup table. This is a common technique for FFT implementations.

Updated with a lookup table.

**Review by philburk — APPROVED** [review]

**Pull request merged**